File: IS/homeworks/frog_leap.py
import copy
import time

def isSolved(board):
    n = int((len(board)-1)/2)
    emptyIndex = n
    if board.index('_') != n:
        return False
    for  counter in range(0,n):
        if board[counter] != '<':
            return False
    return True  
    
def frog_leap(board,emptyIndex,stack):
    if isSolved(board):
        return True
    for _board in available_boards(board):
        if frog_leap(_board,_board.index('_'),stack):
            stack.append(_board)
            return True
    return False
    
def frog_leap_calculation(n):
    start = time.time()
    board = ['>' for x in range(0,n)] + ['_'] + ['<' for x in range(0,n)]
    stack = []
    frog_leap(board,n,stack)
    print(len(stack))
    print(board)
    for i in range(0,len(stack)):
        print(stack.pop(-1))
    end = time.time()
    print(end-start)

def available_boards(board):
    emptyIndex = board.index('_')
    possible_moves = [emptyIndex+x for x in range(-2,3) if emptyIndex+x>=0 and emptyIndex+x<=len(board)-1 and x!=0]
    possible_boards = []
    for addend in possible_moves:
        if (addend<emptyIndex and board[addend] == '>') or (addend>emptyIndex and board[addend] == '<'):
                new_board = copy.copy(board)
                new_board[addend],new_board[emptyIndex] = new_board[emptyIndex],new_board[addend]
                possible_boards.append(new_board)
    return possible_boards

frog_leap_calculation(20)
# RL_RR

File: IS/homeworks/frog_leap.cpp
#include <iostream>
#include <vector>
#include <ctime>

bool isSolved(std::vector<char>& board) {
    int n = (board.size() - 1) / 2;
    int emptyIndex = n;
    if (board[emptyIndex] != '_') {
        return false;
    }
    for (int i = 0; i < n; i++) {
        if (board[i] != '<') {
            return false;
        }
    }
    return true;
}

bool frogLeap(std::vector<char>& board, int emptyIndex, std::vector<std::vector<char>>& stack) {
    if (isSolved(board)) {
        return true;
    }
    std::vector<std::vector<char>> availableBoards = availableBoards(board);
    for (const auto& _board : availableBoards) {
        if (frogLeap(_board, _board[emptyIndex], stack)) {
            stack.push_back(_board);
            return true;
        }
    }
    return false;
}

std::vector<std::vector<char>> availableBoards(std::vector<char>& board) {
    int emptyIndex = 0;
    while (board[emptyIndex] != '_') {
        emptyIndex++;
    }
    std::vector<int> possibleMoves;
    for (int x = -2; x <= 2; x++) {
        int addend = emptyIndex + x;
        if (addend >= 0 && addend < board.size() && x != 0) {
            possibleMoves.push_back(addend);
        }
    }

    std::vector<std::vector<char>> possibleBoards;
    for (int addend : possibleMoves) {
        if ((addend < emptyIndex && board[addend] == '>') || (addend > emptyIndex && board[addend] == '<')) {
            std::vector<char> newBoard = board;
            std::swap(newBoard[addend], newBoard[emptyIndex]);
            possibleBoards.push_back(newBoard);
        }
    }
    return possibleBoards;
}

void frogLeapCalculation(int n) {
    std::clock_t start = std::clock();
    std::vector<char> board;
    for (int i = 0; i < n; i++) {
        board.push_back('>');
    }
    board.push_back('_');
    for (int i = 0; i < n; i++) {
        board.push_back('<');
    }
    std::vector<std::vector<char>> stack;
    frogLeap(board, n, stack);
    std::cout << stack.size() << std::endl;
    std::cout << "RL_RR" << std::endl;
    for (int i = stack.size() - 1; i >= 0; i--) {
        const auto& currentBoard = stack[i];
        for (char c : currentBoard) {
            std::cout << c;
        }
        std::cout << std::endl;
    }
    std::clock_t end = std::clock();
    double elapsedSeconds = static_cast<double>(end - start) / CLOCKS_PER_SEC;
    std::cout << elapsedSeconds << " seconds" << std::endl;
}

int main() {
    frogLeapCalculation(20);
    return 0;
}
